src/VectorBundle.jl

```diff
@@ -0,0 +1,221 @@
+    return v
+end
+
+"""
+    zero_vector(M::Manifold, VS::VectorSpaceType, x)
+
+Compute the zero vector from the vector space of type `VS` at point `x`
+from manifold `M`.
+"""
+function zero_vector(M::Manifold, VS::VectorSpaceType, x)
+    v = similar_result(M, VS, zero_vector, x)
+    zero_vector!(M, VS, v, x)
+    return v
+end
+
+"""
+    project_vector!(M::Manifold, VS::VectorSpaceType, v, x, w)
+
+Project vector `w` from the vector space of type `VS` at point `x`
+and save the result to `v`.
+"""
+function project_vector!(M::Manifold, VS::VectorSpaceType, v, x, w)
+    error("project_vector! not implemented for vector space manifold $(typeof(M)), vector space of type $(typeof(VS)), output vector of type $(typeof(v)) and input vector at point $(typeof(x)) with type of w $(typeof(w)).")
+end
+
+"""
+    inner(M::Manifold, VS::VectorSpaceType, x, v, w)
+
+Inner product of vectors `v` and `w` from the vector space of type `VS`
+at point `x` from manifold `M`.
+"""
+function inner(M::Manifold, VS::VectorSpaceType, x, v, w)
+    error("inner not defined for manifold of type $(typeof(M)), vector space of type $(typeof(VS)), point of type $(typeof(x)) and vectors of types $(typeof(v)) and $(typeof(w)).")
+end
+
+function inner(M::Manifold, VS::TCoTSpaceType, x, v, w)
+    return inner(M, x, v, w)
+end
+
+function project_vector!(M::Manifold, VS::TCoTSpaceType, v, x, w)
+    project_tangent!(M, v, x, w)
+    return v
+end
+
+"""
+    distance(M::Manifold, VS::VectorSpaceType, x, y)
+"""
+function distance(M::Manifold, VS::VectorSpaceType, x, y)
+    error("distance not defined for manifold of type $(typeof(M)), vector space of type $(typeof(VS)) and vectors of types $(typeof(x)) and $(typeof(y)).")
+end
+
+distance(M::Manifold, VS::TCoTSpaceType, x, y) = norm(x-y)
+
+
+"""
+    norm(M::Manifold, VS::VectorSpaceType, x, v)
+
+Norm of the vector `v` from the vector space of type `VS` at point `x`
+from manifold `M`.
+"""
+function norm(M::Manifold, VS::VectorSpaceType, x, v)
+    error("norm not defined for manifold of type $(typeof(M)), vector space of type $(typeof(VS)) and vectors of types $(typeof(x)) and $(typeof(v)).")
+end
+
+norm(M::Manifold, VS::TCoTSpaceType, x, v) = norm(v)
+
+"""
+    VectorBundle(M::Manifold, type::VectorSpaceType)
+
+Vector bundle on manifold `M` of type `type`.
+"""
+struct VectorBundle{TVS<:VectorSpaceType, TM<:Manifold} <: Manifold
+    type::TVS
+    M::TM
+end
+
+function representation_size(M::VectorBundle)
+    len_manifold = prod(representation_size(M.M))
+    len_vs = prod(representation_size(M.M, M.type))
+    return (len_manifold + len_vs,)
+end
+
+TangentBundle(M::Manifold) = VectorBundle(TangentSpace, M)
+CotangentBundle(M::Manifold) = VectorBundle(CotangentSpace, M)
+
+"""
+    bundle_projection(M::VectorBundle, x::ProductRepr)
+
+Bundle projection of point `x` from the vector bundle `M`.
+"""
+function bundle_projection(M::VectorBundle, x::ProductRepr)
+    return x.parts[1]
```

Comment:
I don't really understand from the docstring what this is. It's a projection of the bundle to what? What does the point represent? Also, since it basically does the same thing as `submanifold_component`, perhaps it should call that in case it's specialized for some reason??

Reply:
It's a projection from the bundle to the manifold it is defined on: https://en.wikipedia.org/wiki/Vector_bundle . I guess it can just call `submanifold_component`, right.

Reply:
I think using the phrase "projection from the bundle to the manifold" in the docstring would be more descriptive.